### ISSUE TITLE
PathBindable vs QueryStringBindable

### ISSUE THREAD

**Issue by PierreAdam** [playframework/playframework#11198]
Hi there !

Currently, when you make an object `PathBindable` it's impossible to also make it `QueryStringBindable` at the same time.
The reason being that the method `String unbind(String key);` is utilized by both `PathBindable` and `QueryStringBindable` but in two very different way.
The first expect only the value to add to the path which the second expect a query string syntaxe such as `key1=value1`.
This is a bit weird like for example when you're making entities or models that should be able to be bindable in both the path and the query string.

The easiest way to solve this would be to rename one of the unbind method from one of the interface.
An other solution would be to have a single interface making sure that everything that can be bind to a path can also be bind to a query string.

**Comment by mkurz** [comment]
> The easiest way to solve this would be to rename one of the unbind method from one of the interface.

Yeah, I think that would be the prefered solution. However, we should deprecate the old method names and still support them to not break existing apps.
Do you want to provide a pull request?

**Comment by takapi327** [comment]
@mkurz 
Hi there !

The same seems to be true for PathBindable and QueryStringBindable defined in Scala.
When creating an object that inherits both PathBindable and QueryStringBindable, the unbind and javascriptUnbind and transform methods conflict.
This will cause conflicts even if these three methods are defined under different names as deprecated.

In Scala, would it be better to generate an implicit function instead of generating objects that inherit from QueryStringBindable and PathBindable, as in the sample code in the code?

I couldn't think of a good way to fix QueryStringBindable and PathBindable.
Any ideas?

```console
[info] compiling 15 Scala sources to /Users/takapi327/Development/play-app/sample-ssr-api/target/scala-2.13/classes ...
[error] /Users/takapi327/Development/play-app/sample-ssr-api/app/mvc/util/Binder.scala:8:9: class UserBinder inherits conflicting members:
[error]   def unbind(key: String, value: mvc.util.User): String (defined in trait PathBindable) and
[error]   def unbind(key: String, value: mvc.util.User): String (defined in trait QueryStringBindable)
[error]   (note: this can be resolved by declaring an `override` in class UserBinder.);
[error]  other members with override errors are: javascriptUnbind, transform
[error]   class UserBinder(implicit intBinder: PathBindable[Int], stringBinder: QueryStringBindable[String], intQueryBinder: QueryStringBindable[Int])
[error]         ^
[error] one error found
[error] (Compile / compileIncremental) Compilation failed
[error] Total time: 3 s, completed 2022/04/12 0:21:58

```

**Comment by mkurz** [comment]
@takapi327 If deprecation does not work, we could break compatibility here, rename all the methods (e.g. `bind` becomes `bindQuery` for QueryStringBindable and also `bindPath` for the PathBindable). We can add a note for this in the migration notes for Play 2.9. IMHO that would work, because for major release we don't promise compatibility.

BTW: You work for [Nextbeat](https://www.nextbeat.co.jp/) in Japan? Do you know Tomokazu Uehara? He wrote us an e-mail a couple of month ago, I did reply twice but he never answered...

**Comment by takapi327** [comment]
@mkurz 
I see, I can break compatibility by making a note in the Play 2.9 migration notes!

Create a pull request with this improved method.

I learned a lot. Thanks!

> BTW: You work for [Nextbeat](https://www.nextbeat.co.jp/) in Japan? Do you know Tomokazu Uehara? He wrote us an e-mail a couple of month ago, I did reply twice but he never answered...

I am sorry. I will check on my end.

**Comment by mkurz** [comment]
I was looking into the PR #11244 which was supposed to fix that: However, just renaming the methods will not bind the path and the query string at the same time.
If there is a path fragment to bind, Play calls following `fromPath` method with the implicit path bindable, completely ignoring the query string. The same for the query string bindable, the other way round.
https://github.com/playframework/playframework/blob/58ee8806c675f95eb5ce89662d3e1a5a85d62b6f/core/play/src/main/scala/play/core/routing/GeneratedRouter.scala#L62-L88

I played around with the patch from #11244 and this is what the code that Play generates for some routes I was testing:

```scala
    // @LINE:8
    case controllers_HomeController_user1_route(params@_) =>
      call(params.fromPath[models.User]("user", None), params.fromQuery[models.User]("foo", None)) { (user, foo) =>
        controllers_HomeController_user1_invoker.call(HomeController_0.user(user, foo))
      }
````

which gets generated here:
https://github.com/playframework/playframework/blob/58ee8806c675f95eb5ce89662d3e1a5a85d62b6f/dev-mode/routes-compiler/src/main/scala/play/routes/compiler/templates/package.scala#L89-L90

This is the route I was using:
```
GET     /user/:user            controllers.HomeController.user(user: models.User, foo: models.User)
```

The `user` param does not bind any query string because the query string binder is not called. Therefore I added a `foo` param but now this time of course `foo` only binds the query string params, ignoring the path fragment. (of course, because Play matches the `:user` name with the `user` param and only then binds the path.

I think the goal here was to set up a route like
```
GET     /user/:user            controllers.HomeController.user(user: models.User)
```
and have user bind the path _and_ the query string to the `user` instance (?)

If so, I think the only thing that we need to do is to also pass the queryparams to the path binders `bind` method, so 
https://github.com/playframework/playframework/blob/58ee8806c675f95eb5ce89662d3e1a5a85d62b6f/core/play/src/main/scala/play/api/mvc/Binders.scala#L166

becomes
```scala
 def bind(key: String, value: String, queryParams: Map[String, Seq[String]]): Either[String, A]
```

That should be doable, we could even keep the original `bind` for one release to not break existing apps.

**Comment by mkurz** [comment]
@PierreAdam What was the real world use case that motivated you do open this issue?

Do you want to be able to also have access to the query string params when using a path binder?
So when having a route like
```
GET     /user/:user            controllers.HomeController.user(user: models.User)
```
and calling it with
```
/user/2?age=39
```
so you can access the `age` query string param in the Pathbinders `bind` method??
If so we should go with the solution I proposed at the end of my previous and just add the `queryParams` map to the PathBinder's bind method.

Or do you want to be able to use the _same_ binder for routes like:
```
# called with /user/2
GET     /user/:user   controllers.HomeController.user(user: models.User)

# called with /user?id=2
GET     /user            controllers.HomeController.user(user: models.User)
```
if so, that requires much deeper changes in how we handle binding and I am not sure we will implement that anytime soon.

BTW:
Taking the first approach further people might even want to do things like:
```
GET     /branch/:branch/user/:user   controllers.HomeController.user(user: models.User)
```
where a binder is able to handle both `:branch` and `:user` path fragments and is also able to access querie strings and then spit out the `User` instance... I don't think this makes sense right now, I just wanted to mention that.
